Support composite foreign keys via migration helpers

fatkodima opened the pull request:
Closes #47593.

```ruby
# Assuming "carts" table has "(shop_id, id)" as a primary key.

add_foreign_key(:orders, :carts, primary_key: [:shop_id, :id])
# or
add_foreign_key(:orders, :carts, column: [:cart_shop_id, :cart_id])

remove_foreign_key(:orders, :carts, primary_key: [:shop_id, :id])
foreign_key_exists?(:orders, :carts, primary_key: [:shop_id, :id])
```

We [discussed](https://github.com/rails/rails/issues/47593#issuecomment-1458359213) 3 possible ways of using `add_foreign_key`.

The first one (`add_foreign_key :brochures, :cars`) (no `:column` and `:primary_key` options) turned out to be complex to implement, as it introduced a lot of kinda ugly changes in many places because we needed to properly generate these column names (consider configured table name prefixes and suffixes, pass a connection around to be able to query a primary key for the `to_table`. Some classes, like `ForeignKeyDefinition` does not have access to the connection, and so we needed to calculate these columns at several places differently and pass there). 

And the biggest problem is when we need to add a foreign key for a self referencing tables, like https://github.com/rails/rails/blob/1df6ad0ecc8bdefa233eb388c38d1ff845862428/activerecord/test/cases/migration/references_foreign_key_test.rb#L221-L225 That would produce very not elegant code.

So I made that it is required to pass at least one of `:column` or `:primary_key` options, as it is easy to infer one from the other.

cc @eileencodes @nvasilevski

fatkodima committed:
Support composite foreign keys via migration helpers

eileencodes commented on activerecord/CHANGELOG.md:
````diff
@@ -1,3 +1,20 @@
+*   Support composite foreign keys via migration helpers.
+
+    ```ruby
+    # Assuming "carts" table has "(shop_id, id)" as a primary key.
````
I think we are actually going to discourage using `id` as one of the ids in the composite key when we write the docs so can we change this to something like `(shop_id, oid)`?

eileencodes commented on activerecord/lib/active_record/connection_adapters/abstract/schema_statements.rb:
```diff
@@ -1099,6 +1099,17 @@ def foreign_keys(table_name)
+      #
+      #   ALTER TABLE "orders" ADD CONSTRAINT fk_rails_6f5e4cb3a4 FOREIGN KEY ("cart_shop_id", "cart_id") REFERENCES "carts" ("shop_id", "id")
+      #
+      # Note: At least one of +column+ or +primary_key+ options must be provided.
```
I'm not super clear on why we need to support both, can we support just primary_key and simplify the code a bit?

eileencodes reviewed (changes requested):
Hey @fatkodima thanks for working on this! I've left some comments.

fatkodima commented:
@eileencodes Thanks for the review, addressed the feedback.

I initially added support for both `:primary_key` and `:column` to be able to specify either one and the other will be inferred. Small convenience. But yeah, changed to require passing a `:primary_key` manually, if the `:column` is passed as an array.

rafaelfranca committed:
Make branches explict

We only set the column to a single column if the primary key is a single
value, so move that code to an explicit else.

rafaelfranca committed:
Avoid default value

This method is internal so we can just always pass the argument.
